apache/trafficcontrol
# Allow IPv6 Only Caches, including monitoring and routing

## What does this PR (Pull Request) do?


- [x] This PR fixes #2735

It is also an update to PR #4012 so that one can be closed


## Which Traffic Control components are affected by this PR?


- Documentation
- Traffic Monitor
- Traffic Ops
- Traffic Portal
- Traffic Router

## What is the best way to verify this PR?


- Update CDN in a Box Docker to include IPv6 or test in an environment that has IPv6.  
- Verify that the new IPv4 and IPv6 columns are added to the TM UI and are showing the correct boolean for availability for IPv4 (since IPv4 only is the default) and that it is polling successfully. 
- Update the TM config file to include
"cache_polling_protocol": "both",
"peer_polling_protocol": "both"

- Restart TM with the new config. Verify that it is successfully polling availability information for both IPv4 and IPv6. Verify that servers with IPv4 are still successfully polled. Verify that any issues are being reported correctly for both IPv4 and IPv6.
- Verify that TP allows servers to have an IPv6 only address, but does require either IPv4, IPv6, or both to be defined.
- Create a new server object or update an old server object to have only an IPv6 address.  Verify that this its availability was updated in TM.
- Run a curl command with -4 or -6 to verify that the IPv4 and IPv6 routing both work.  
- In TP, uncheck the IsServiceAddress box for IPv4 to disable this address.  Perform a snapshot.
- Verify that IPv4 is marked as unavailable in TM, but that the Cache as a whole is marked as available still.
- Verify that a curl with -6 still works but a curl with -4 returns a 503 status.
- Repeat these steps but uncheck the IsServiceAddress for IPv6 and verify that -4 is functional and -6 returns a 503.
- Update IPv4 address to "127.0.0.0" (or any valid IPv4 address that will get an error when TM requests status) and verify that it gets an error in TM and has the IPv4 marked unavailable as expected.
- Update IPv6 address to "::1" (or any valid IPv6 address that will get an error when TM requests status) and verify that it gets an error in TM and has the IPv6 marked unavailable as expected.

## The following criteria are ALL met by this PR




## Additional Information
